### PR TITLE
Handling the ContinuedEvent

### PR DESCRIPTION
**Pull request by asimgunes**
This update implements handling mechanism for "resume" gdb signal and redirect the "resume" signal as "ContinuedEvent" as described in the Debug Adapter Protocol.

This implementation is related to the issue #332

### CODE REVIEW

**Comment by jonahgraham** [comment]
@asimgunes this looks promising - it over sends ContinuedEvent, which is not strictly a bug, but may be undesired. From the [DAP docs](https://microsoft.github.io/debug-adapter-protocol/specification#Events_Continued:~:text=Please%20note%3A%20a,that%20implied%20this.):

> Please note: a debug adapter is not expected to send this event in response to a request that implies that execution continues, e.g. launch or continue.
> 
> It is only necessary to send a continued event if there was no previous request that implied this.

See [here](https://github.com/Microsoft/vscode/blob/a5f52063e4622bc318d9c550a682dc5b35ef7f33/src/vs/workbench/contrib/debug/browser/rawDebugSession.ts#L334) that VSCode already sends a simulated continued event on operations like next. 

We could avoid sending these unnecessary ContinuedEvents to avoid unnecessary updates to the UI and related code.

The flip-side of this issue is that we aren't testing that we are sending continued events when no previous request implies one. To achieve this, instead of doing `dc.continueRequest` to continue, send a CLI command for `continue` either by using `cdt-gdb-tests/executeCommand` as you did previously [here](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/590fea58cfb2ee4984d766cd1b2140738d3ff110/src/integration-tests/launchWithEnvironment.spec.ts#L43-L45), or by sending an evaluate request with the `>` prefix to make it a GDB command like [here](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/b593b5798174dcd8d68677086b5847b94d9d61e7/src/integration-tests/evaluate.spec.ts#L97-L102)

File: src/integration-tests/multithread.spec.ts [hunk 1]
```diff
@@ -182,4 +182,58 @@ describe('multithread', async function () {
...
+                    ? {
+                          threadId: thread.id,
+                          allThreadsContinued: false,
+                      }
+                    : { threadId: 1, allThreadsContinued: true }
+            );
+
+            // if gdbNonStop=false stop the thread for next iteration
+            if (!gdbNonStop) {
+                await dc.pauseRequest({ threadId: thread.id });
```
Comment (jonahgraham) [inline src/integration-tests/multithread.spec.ts:234]: This won't work on Windows in non-async case. see https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/ffeb53a09495ba20968179303ecb35168415cd1f/src/integration-tests/breakpoints.spec.ts#L103-L106

**Review by jonahgraham — CHANGES_REQUESTED** [review]
@asimgunes I have completed a review

File: src/integration-tests/multithread.spec.ts (hunk 1 above)
Comment (asimgunes) [reply src/integration-tests/multithread.spec.ts]: I made a similar implementation, thanks for the suggestion.

**Review by jonahgraham — CHANGES_REQUESTED** [review]
This LGTM, apart from thread id == 1, see line comments

File: src/integration-tests/multithread.spec.ts [hunk 2]
```diff
@@ -182,4 +184,98 @@ describe('multithread', async function () {
...
+        // Send continue to thread 2
+        dc.send('cdt-gdb-tests/executeCommand', {
+            command: '-exec-continue --thread 2',
+        });
+
+        const event = await dc.waitForEvent('continued');
+
+        // In allThreadsContinued:true case we are expecting threadId as '1' no matter which thread is continued
+        assert.deepEqual(event.body, {
+            threadId: 1,
```
[the remark this answers is not in the document]
Comment (asimgunes) [reply src/integration-tests/multithread.spec.ts]: Hi Jonah, you are right it is not going to be '1' in every case, and I made the change due to your suggestion. However, I kept sending 1 in any unexpected case like if threads list is empty. We may also consider failing or skipping to send the continued signal if threads list is empty; but, I couldn't be sure which approach would be better. Could you please re-check the implementation, and let me know if you have any other concerns.

File: src/gdb/GDBDebugSessionBase.ts [hunk 3]
```diff
@@ -1601,6 +1602,37 @@ export abstract class GDBDebugSessionBase extends LoggingDebugSession {
...
+                !!resultData['thread-id'] && resultData['thread-id'] === 'all'
+            );
+        };
+
+        const isAllThreadsContinued = getAllThreadsContinued(result);
+        if (isAllThreadsContinued) {
+            // If all threads continued, then the value of the 'thread-id' is 'all',
+            // hence, there isn't a thread id number. We are sending the thread id as '1'
+            // along with the allThreadsContinued=true information.
+            return this.sendContinuedEvent(1, true);
```
[the remark this answers is not in the document]
Comment (asimgunes) [reply src/gdb/GDBDebugSessionBase.ts]: Hi Jonah, you are right it is not going to be '1' in every case, and I made the change due to your suggestion. However, I kept sending 1 in any unexpected case like if threads list is empty. We may also consider failing or skipping to send the continued signal if threads list is empty; but, I couldn't be sure which approach would be better. Could you please re-check the implementation, and let me know if you have any other concerns.

**Comment by asimgunes** [comment]
Hi @jonahgraham ,

I made the suggested changes, could you please review again when you are suitable?

Could you please check the update when you are suitable?

File: src/gdb/GDBDebugSessionBase.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonahgraham) [reply src/gdb/GDBDebugSessionBase.ts]: LGTM

**Review by jonahgraham — APPROVED** [review]

**Comment by jonahgraham** [comment]
The version with this fix is published to NPM https://www.npmjs.com/package/cdt-gdb-adapter/v/1.0.1-next.20240927165134.44fda62.0

**Pull request merged**